### PR TITLE
adds conversion function from C to Rust object for CSlot and CSlotValue.

### PR DESCRIPTION
**Pull request by anthonyray**
Hi guys, 
I needed conversion methods for Slots. The conversions from C to Rust object where missing for Slot objects. 

I also took the initiative to introduce the very practical trait : `AsRust`, and thus diverge from implementing the `From` trait. 

I also took the liberty to introduce `failure` as a dependency, so that we are consistent across crates. 

Let me know if I need to fix somethin.

### CODE REVIEW

File: ffi/ffi-macros/src/ontology.rs [hunk 1]
```diff
@@ -141,6 +142,21 @@ impl From<Slot> for CSlot {
...
+impl AsRust<Slot> for CSlot {
+    fn as_rust(&self) -> Fallible<Slot> {
+        Ok(
+            Slot {
+                raw_value: create_rust_string_from!(self.raw_value),
+                value: self.value.as_rust()?,
+                range: (self.range_start as usize..self.range_end as usize),
+                entity: create_rust_string_from!(self.entity),
+                slot_name: create_rust_string_from!(self.slot_name),
+                confidence_score: Some(self.confidence_score),
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: It should rather be:
```rust
confidence_score: if self.confidence_score < 0 { None } else { Some(self.confidence_score) } 
```

Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: It should rather be:
```rust
confidence_score: if self.confidence_score < 0 { None } else { Some(self.confidence_score) } 
```

File: ffi/ffi-macros/src/ontology.rs [hunk 2]
```diff
@@ -461,6 +477,111 @@ impl From<SlotValue> for CSlotValue {
...
+                        SNIPS_GRAIN::SNIPS_GRAIN_DAY => Grain::Day,
+                        SNIPS_GRAIN::SNIPS_GRAIN_HOUR => Grain::Hour,
+                        SNIPS_GRAIN::SNIPS_GRAIN_MINUTE => Grain::Minute,
+                        SNIPS_GRAIN::SNIPS_GRAIN_SECOND => Grain::Second
+                    },
+                    precision: match c_instant_time_value.precision {
+                        SNIPS_PRECISION::SNIPS_PRECISION_APPROXIMATE => Precision::Approximate,
+                        SNIPS_PRECISION::SNIPS_PRECISION_EXACT => Precision::Exact
+                    },
+                };
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: Can't we implement `AsRust` for `CInstantTimeValue` ?

File: ffi/ffi-macros/src/ontology.rs [hunk 3]
```diff
@@ -461,6 +477,111 @@ impl From<SlotValue> for CSlotValue {
...
+                Ok(SlotValue::InstantTime(instant_time_value))
+            }
+            SNIPS_SLOT_VALUE_TYPE::SNIPS_SLOT_VALUE_TYPE_TIMEINTERVAL => {
+                let c_time_interval_value = unsafe { &*(self.value as *const CTimeIntervalValue) };
+                let time_interval_value = TimeIntervalValue {
+                    from: create_optional_rust_string_from!(c_time_interval_value.from),
+                    to: create_optional_rust_string_from!(c_time_interval_value.to),
+                };
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: Can't we implement `AsRust` for `CTimeIntervalValue` ?

File: ffi/ffi-macros/src/ontology.rs [hunk 4]
```diff
@@ -461,6 +477,111 @@ impl From<SlotValue> for CSlotValue {
...
+
+                let amount_of_money_value = AmountOfMoneyValue {
+                    value: c_amount_of_money_value.value,
+                    precision: match c_amount_of_money_value.precision {
+                        SNIPS_PRECISION::SNIPS_PRECISION_EXACT => Precision::Exact,
+                        SNIPS_PRECISION::SNIPS_PRECISION_APPROXIMATE => Precision::Approximate,
+                    },
+                    unit: create_optional_rust_string_from!(c_amount_of_money_value.unit),
+                };
+                Ok(SlotValue::AmountOfMoney(amount_of_money_value))
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: Same comment

File: ffi/ffi-macros/src/ontology.rs [hunk 5]
```diff
@@ -461,6 +477,111 @@ impl From<SlotValue> for CSlotValue {
...
+                    days: c_duration_value.days,
+                    hours: c_duration_value.hours,
+                    minutes: c_duration_value.minutes,
+                    seconds: c_duration_value.seconds,
+                    precision: match c_duration_value.precision {
+                        SNIPS_PRECISION::SNIPS_PRECISION_APPROXIMATE => Precision::Approximate,
+                        SNIPS_PRECISION::SNIPS_PRECISION_EXACT => Precision::Exact,
+                    },
+                };
+                Ok(SlotValue::Duration(duration_value))
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: same comment

File: ffi/ffi-macros/src/ontology.rs [hunk 6]
```diff
@@ -461,6 +477,111 @@ impl From<SlotValue> for CSlotValue {
...
+                Ok(SlotValue::MusicArtist(StringValue {
+                    value: create_rust_string_from!(self.value as *const libc::c_char)
+                }))
+            }
+            SNIPS_SLOT_VALUE_TYPE::SNIPS_SLOT_VALUE_TYPE_MUSICTRACK => {
+                Ok(SlotValue::MusicTrack(StringValue {
+                    value: create_rust_string_from!(self.value as *const libc::c_char)
+                }))
+            }
+            _ => bail!("The provided slot value type doesn't exists. Cannot perform conversion to Rust object ")
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: It would be great to include `self.value_type` in the error message.

**Comment by adrienball** [comment]
Now that you have implemented `AsRust`, you can add some round trip unit tests in rust.

File: ffi/ffi-macros/src/ontology.rs (hunk 1 above)
Comment (anthonyray) [reply ffi/ffi-macros/src/ontology.rs]: Oh yesss, thanks for catching this one 🙊 !

File: ffi/ffi-macros/src/ontology.rs (hunk 2 above)
Comment (anthonyray) [reply ffi/ffi-macros/src/ontology.rs]: Sure thing, it would be cleaner.

File: ffi/ffi-macros/src/ontology.rs [hunk 7]
```diff
@@ -541,3 +721,174 @@ impl Drop for CSlotValue {
...
+    pub fn round_trip_test<T, U>(input: T)
+        where
+            T: Clone + PartialEq + std::fmt::Debug,
+            U: From<T> + AsRust<T>,
+    {
+        let c = U::from(input.clone());
+
+        let result = c.as_rust().expect("could not convert back to rust");
+        assert_eq!(result, input);
+    }
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: 👍

File: ffi/ffi-macros/src/ontology.rs [hunk 8]
```diff
@@ -541,3 +721,174 @@ impl Drop for CSlotValue {
...
+    #[test]
+    fn round_trip_c_slot() {
+        round_trip_test::<_, CSlot>(Slot {
+            raw_value: "raw_value".to_string(),
+            value: SlotValue::Custom("slot_value".to_string().into()),
+            range: 0..1,
+            entity: "entity".to_string(),
+            slot_name: "slot_name".to_string(),
+            confidence_score: Some(0.5),
+        });
```
Comment (adrienball) [inline ffi/ffi-macros/src/ontology.rs]: Could you add a test with a `confidence_score: None` ?
